Solutions/Corvus.Leasing.Abstractions/Corvus/Leasing/Lease.cs

```diff
@@ -33,7 +33,7 @@ protected Lease(ILeaseProvider leaseProvider, LeasePolicy leasePolicy, string id
         /// <summary>
         /// Gets or sets the lease id.
         /// </summary>
-        public string Id { get; protected set; }
+        public string? Id { get; protected set; }
```

Comment:
Why is this nullable? The constructor requires a non-null `id` argument, so to construct a lease you need a non-null id.

Looking at the code I see that the `AzureLeaseProvider.cs` unsets this, but apparently this is as a slightly indirect means of putting the lease into a "no longer acquired" state—it seems to be being done in order to ensure `HasLease` becomes `false`. But is it really necessary to make the `Id` become null at the same time?

Reply:
Yes. This is a tricky one, and I discussed with @mwadams at the point I encountered it. I agree that in an ideal world it would be non-nullable, but the outcome of the decision was that making that change had the potential to break clients who are relying on the assumption that a null Id means that a lease is no longer usable. 

The two options would be:
- not change the Id when the lease is released, and add a flag to indicate that the release has taken place.
- set the Id to Guid.Empty when the release happens.

I've noticed that whilst the `AzureLease` subclass has some specific behaviour around clearing the Id when a lease is released, the InMemoryLeaseProvider doesn't, which means that when using that provider, `HasLease` will only return true once a least has expired. This also seems incorrect.

So, given that, my feeling is that we should:
- make `Lease.Id` non-nullable
- add a new boolean property, `Released` which will be set in the `Lease.ReleaseAsync` method once `this.LeaseProvider.ReleaseAsync` has been called
- change `Lease.HasLease` to use the new flag instead of null-checking the Id.

Thoughts? (cc @mwadams )

Reply:
Agreed. If we are going to break it, let's break it like that.

Reply:
Released is readonly, I presume.

Reply:
Changes made.